# Examen A Módulo 1 - Optimización.

**Lea cuidadosamente las siguientes indicaciones antes de comenzar el examen:**
- Para resolver el examen edite este mismo archivo y renómbrelo de la siguiente manera: *Examen1_ApellidoNombre*, donde *ApellidoNombre* corresponde a su apellido paterno con la inicial en mayúscula, seguido de su primer nombre con la inicial en mayúscula **sin tildes**. Por ejemplo, en mi caso el nombre del archivo sería *Examen1_MoralesGabriel*.
- Resuelva los puntos en el espacio provisto. Si requiere agregar más celdas para código o escritura, hágalo (`Esc`,`B`) o (`Esc`,`A`).
- Recuerde que también se está evaluando su capacidad de interpretar los resultados. Escriba sus interpretaciones/conclusiones en celdas utilizando *Markdown*.
- El formato de presentación de todo el examen en general debe ser adecuado. Use tamaños de letra, colores, etiquetas, etcétera.
- No se resuelven dudas de ningún tipo por los profesores de la asignatura. Por favor, absténgase de preguntar.
- Revise bien el tipo de examen. Sólo se le calificará si realiza el tipo de examen asignado.

## 1. Optimización de funciones con `Sympy`

Funciones permitidas: subs, solve, var, expresiones matemáticas


Un cine cobra 86 pesos la entrada estándar y 60 pesos la entrada a estudiantes. Una sala con capacidad para 120 personas se ocupó al 70\%. Los ingresos para esa función fueron de 6,600 MXN. ¿Cuántos estudiantes entraron?

In [7]:
import sympy as sp

x = sp.var('x')
y = sp.var('y')

eq1 = sp.Eq(x + y, 0.7 * 120)
eq2 = sp.Eq(60 * x + 86 * y, 6600)

solucion = sp.solve((eq1, eq2), (x, y))

estudiantes = solucion[x]
estudiantes


24.0000000000000

Esto nos quiere decir que fueron 24 estudiantes los que ingresaron a la funcion y compraron entradas preferenciales, mientras que 60 personas compraron entradas a precio normal. Por lo tanto los ingresos por estudiantes serían 24×60=1440 pesos, y los ingresos por entradas estándar serían 60×86=5160 pesos y la suma de esto da 6600 pesos.


El cine ofrece combos y opciones individuales. El combo 'amigos' tiene un costo de 156 MXN y consiste de 2 refrescos y unas palomitas. Existe también el combo 'nachos' que tiene un costo de 112 MXN y consiste de un refresco y una orden de nachos. Por separado el refresco tiene un costo de 60 MXN, las palomitas 58 MXN y los nachos 72 MXN. Se obtuvo un total de 6450 de ingresos en dulcería:

- 38 combos vendidos
- 70 refrescos vendidos
- 33 órdenes de palomitas grandes vendidas
- 16 órdenes de nachos vendidas

¿Cuántas personas compraron palomitas sueltas?

combo_amigos = 156

combo_nachos = 112

refresco = 60

palomitas = 58

nachos = 72

venta_total = 6450




In [17]:
import sympy as sp

x = sp.var('x')
y = sp.var('y')
z = sp.var('z')

combo_amigos = 156
combo_nachos = 112
refresco = 60
palomitas = 58
nachos = 72
venta_total = 6450

eq1 = sp.Eq(38 * 156 + y * 60 + 33 * 58 + z * 72, 6450)
eq2 = sp.Eq(70, 38 * 2 + y)
eq3 = sp.Eq(z, 16)

solucion = sp.solve((eq1, eq2, eq3), (x, y, z))

solucion



[]

## 2. Programación lineal

Una empresa produce costales de 3 tipos de fertilizantes para diferentes aplicaciones utilizando cuantro componentes químicos: azufre, potasio, fósforo y nitrógeno. La información de cada tonelada de fertilizante es:


<table>
  <tr>
      <th>Fertilizante</th>
      <th>Fósforo</th>
      <th>Azufre</th>
      <th>Potasio</th>
      <th>Nitrógeno</th>
      <th>Precio de venta </th>
  </tr>
  <tr>
      <td>A</td>
      <td>3 kg</td>
      <td>5 kg</td>
      <td>4 kg</td>
      <td>2 kg </td>
      <td>\$40000</td>
  </tr>
  <tr>
      <td>B</td>
      <td>4 kg</td>
      <td>3 kg</td>
      <td>2 kg</td>
      <td>1 kg</td>
      <td>\$50000</td>
  </tr>
  <tr>
      <td>C</td>
      <td>5 kg</td>
      <td>2 kg</td>
      <td>3 kg</td>
      <td>4 kg</td>
      <td>\$60000</td>
  </tr>

</table>

La máquina que se usa para mezclar el fertilizante C necesita mantenimiento y por lo tanto sólo puede mezclar 10 toneladas. Además, hay un pedido de 8 toneladas del fertilizante A.

Dado que en almacén se tienen 90 kg de fósforo, 120 kg de nitrógeno, 80 kg de azufre y 95 kg de potasio, ¿cuánto recomiendas que se mezcle de cada fertilizante? (Nota: No se permiten fracciones de toneladas)

90 kg fosforo

120 kg nitrogeno

80 kg de azufre

95 kg de potasio


fosforo = 3 + 4 + 5 $\leq$ 90

Azufre = 5 + 3 + 2 $\leq$ 80

Potasio = 4 + 2 + 3 $\leq$ 95

Nitrogeno = 2 + 1 + 4 $\leq$ 120


# Problema de Programación Lineal para la Producción de Fertilizantes

## Restricciones

1. La máquina que se usa para mezclar el fertilizante C puede mezclar un máximo de 10 toneladas.
2. Hay un pedido de 8 toneladas del fertilizante A.
3. Disponibilidad en almacén:
   - Fósforo: 90 kg
   - Nitrógeno: 120 kg
   - Azufre: 80 kg
   - Potasio: 95 kg

## Variables de Decisión

- \( A \): toneladas del fertilizante A a mezclar
- \( B \): toneladas del fertilizante B a mezclar
- \( C \): toneladas del fertilizante C a mezclar

## Función Objetivo

Maximizar el ingreso total por la venta de fertilizantes:

$[\text{Maximizar } Z = 40000 A + 50000 B + 60000 C ]$

## Restricciones

1. Disponibilidad de fósforo:
   $[
   3A + 4B + 5C \leq 90
   ]$
   
2. Disponibilidad de azufre:
   $[
   5A + 3B + 2C \leq 80
   ]$

3. Disponibilidad de potasio:
   $[
   4A + 2B + 3C \leq 95
   ]$

4. Disponibilidad de nitrógeno:
   $[
   2A + 1B + 4C \leq 120
   ]$

5. Restricción de producción de A:
   $[
   A \geq 8
   ]$

6. Restricción de producción de C:
   $[
   C \leq 10
   ]$

7. No negatividad:
   $[
   A, B, C \geq 0
   ]$

## Solución

Se busca determinar la cantidad de cada fertilizante (A, B y C) a mezclar que maximice los ingresos bajo las restricciones anteriores.


In [35]:
max_fosforo = 90
max_azufre = 80
max_potasio = 95
max_nitrogeno = 120

pedido_A = 8
max_C = 10

mejor_solucion = None
max_ingreso = 0

for B in range(0, 17):
    for C in range(0, max_C + 1):
        A = pedido_A

        fosforo = 3 * A + 4 * B + 5 * C
        azufre = 5 * A + 3 * B + 2 * C
        potasio = 4 * A + 2 * B + 3 * C
        nitrogeno = 2 * A + 1 * B + 4 * C

        if (fosforo <= max_fosforo and
            azufre <= max_azufre and
            potasio <= max_potasio and
            nitrogeno <= max_nitrogeno):

            ingreso = 40000 * A + 50000 * B + 60000 * C

            if ingreso > max_ingreso:
                max_ingreso = ingreso
                mejor_solucion = (A, B, C)

mejor_solucion, max_ingreso

((8, 9, 6), 1130000)

Los resultados sugieren que, dadas las limitaciones de recursos y la necesidad de producir al menos 8 toneladas del Fertilizante A, 9 toneladas del Fertilizante B y limitando la producción de Fertilizante C a 6 toneladas. Esta estrategia maximiza los ingresos, asegurando el cumplimiento de los requisitos de producción y el uso eficiente de los recursos disponibles, ademas aumentando las ganancias significativamente a más de 1'000,000 de pesos.

¿Cuál es el ingreso total?

1'130,000 pesos

¿Cuánto fósforo quedó en almacén?

Fosforo usado=3(8)+4(9)+5(6)

Haciendo las sumas correspondientes y posteriormente restandolo con el total de al final el fosforo restante es de 0

## 3. Regresión lineal

Se quiere predecir la resistencia a la compresión de concreto en función de las proporciones de los componentes en la mezcla (cemento, arena, agua, agregado). La constructora que está realizando el análisis tiene datos históricos donde se registraron cantidades de materia y la resistencia medida:

In [ ]:
import pandas as pd
data = {
    'Cemento (kg)': [350, 400, 300, 450, 380, 410, 320, 370, 390, 360,
                     340, 430, 315, 405, 425, 345, 395, 355, 440, 310,
                     385, 370, 425, 335],
    'Arena (kg)': [600, 550, 700, 500, 580, 540, 650, 590, 570, 610,
                   630, 520, 660, 545, 530, 620, 560, 605, 515, 670,
                   575, 590, 535, 625],
    'Agua (kg)': [180, 160, 200, 170, 175, 165, 185, 172, 170, 190,
                  182, 165, 195, 175, 168, 180, 175, 185, 162, 198,
                  178, 172, 170, 180],
    'Agregado (kg)': [1000, 1100, 950, 1200, 1050, 1150, 980, 1020, 1080, 970,
                      990, 1120, 950, 1060, 1140, 990, 1040, 1010, 1160, 940,
                      1070, 1030, 1110, 1025],
    'Resistencia a la compresión (MPa)': [35, 42, 30, 50, 38, 44, 32, 36, 40, 34,
                                          33, 48, 31, 43, 47, 37, 41, 35, 49, 29,
                                          39, 36, 45, 33]
}
data = pd.DataFrame(data)
data

Cemento (kg)  Arena (kg)  Agua (kg)  Agregado (kg)  \
0            350         600        180           1000   
1            400         550        160           1100   
2            300         700        200            950   
3            450         500        170           1200   
4            380         580        175           1050   
5            410         540        165           1150   
6            320         650        185            980   
7            370         590        172           1020   
8            390         570        170           1080   
9            360         610        190            970   
10           340         630        182            990   
11           430         520        165           1120   
12           315         660        195            950   
13           405         545        175           1060   
14           425         530        168           1140   
15           345         620        180            990   
16           395         560        175           1040   
17           355         605        185           1010   
18           440         515        162           1160   
19           310         670        198            940   
20           385         575        178           1070   
21           370         590        172           1030   
22           425         535        170           1110   
23           335         625        180           1025   

    Resistencia a la compresión (MPa)  
0                                  35  
1                                  42  
2                                  30  
3                                  50  
4                                  38  
5                                  44  
6                                  32  
7                                  36  
8                                  40  
9                                  34  
10                                 33  
11                                 48  
12                                 31  
13                                 43  
14                                 47  
15                                 37  
16                                 41  
17                                 35  
18                                 49  
19                                 29  
20                                 39  
21                                 36  
22                                 45  
23                                 33

Ajusta un modelo utilizando mínimos cuadrados para encontrar una función que pueda predecir la resistencia del concreto.

In [42]:
import numpy as np
import pandas as pd
from scipy.optimize import minimize
from scipy import stats

data = {
    'Cemento (kg)': [350, 400, 300, 450, 380, 410, 320, 370, 390, 360,
                     340, 430, 315, 405, 425, 345, 395, 355, 440, 310,
                     385, 370, 425, 335],
    'Arena (kg)': [600, 550, 700, 500, 580, 540, 650, 590, 570, 610,
                   630, 520, 660, 545, 530, 620, 560, 605, 515, 670,
                   575, 590, 535, 625],
    'Agua (kg)': [180, 160, 200, 170, 175, 165, 185, 172, 170, 190,
                  182, 165, 195, 175, 168, 180, 175, 185, 162, 198,
                  178, 172, 170, 180],
    'Agregado (kg)': [1000, 1100, 950, 1200, 1050, 1150, 980, 1020, 1080, 970,
                      990, 1120, 950, 1060, 1140, 990, 1040, 1010, 1160, 940,
                      1070, 1030, 1110, 1025],
    'Resistencia a la compresión (MPa)': [35, 42, 30, 50, 38, 44, 32, 36, 40, 34,
                                          33, 48, 31, 43, 47, 37, 41, 35, 49, 29,
                                          39, 36, 45, 33]
}
data = pd.DataFrame(data)

X = data[['Cemento (kg)', 'Arena (kg)', 'Agua (kg)', 'Agregado (kg)']].values
y = data['Resistencia a la compresión (MPa)'].values

ones = np.ones((X.shape[0], 1))
X = np.hstack((ones, X))

def costo_funcion(theta, X, y):
    predicciones = X.dot(theta)
    return np.mean((predicciones - y) ** 2)

theta_01 = np.zeros(X.shape[1])

result = minimize(costo_funcion, theta_01, args=(X, y))

theta_estimada = result.x
theta_estimada

array([-5.57589010e+01,  1.37376591e-01,  2.43571125e-02,  1.94850667e-02,
        2.39738351e-02])

In [48]:
predictions = X.dot(theta_estimada)
residual = y - predictions

sigma_squared = np.mean(residual ** 2)

var_theta = sigma_squared * np.linalg.inv(X.T.dot(X))

std_error = np.sqrt(np.diag(var_theta))

t_values = theta_estimada / std_error
p_values = 2 * (1 - stats.t.cdf(np.abs(t_values), df=X.shape[0] - X.shape[1]))

t_values, p_values

(array([-1.66363793,  2.96670803,  0.63903057,  0.35848138,  2.31310047]),
 array([0.11259171, 0.00792418, 0.53043226, 0.72393538, 0.03207882]))

¿Qué efecto tiene cada factor en la resistencia de la mezcla?

## Cemento
* Valor *t*: -1.66
* Valor *p*: 0.11
  
Aunque el cemento es fundamental para la resistencia del concreto, su coeficiente no es significativo. Esto sugiere que, en el rango de datos analizados, los cambios en la cantidad de cemento no tienen un efecto claro sobre la resistencia a la compresión. Esto podría deberse a que otras variables son más influyentes.

## Arena
* Valor *t*: 2.97
* Valor *p*: 0.01

La arena tiene un efecto positivo y significativo en la resistencia a la compresión. Un incremento en la cantidad de arena en la mezcla se asocia con un aumento en la resistencia. Esto indica que optimizar la cantidad de arena podría ser clave para mejorar la calidad del concreto.

## Agua
* Valor *t*: 0.64
* Valor *p*: 0.53

Con base a los resultados el agua no tiene un efecto significativo en la resistencia a la compresión, ya que su valor *p* es bastante alto. Esto indica que, en el rango de datos analizados, la cantidad de agua no influye notablemente en la resistencia, lo que puede ser inesperado, dado que típicamente se espera que el exceso de agua perjudique la resistencia.

## Agregado
* Valor *t*: 0.36
* Valor *p*: 0.72

De igual manera que como ocurrio en el agua, el agregado tampoco muestra un efecto significativo en la resistencia a la compresión. Esto podría indicar que, en el rango de datos, las variaciones en la cantidad de agregado no están afectando notablemente la resistencia del concreto.

¿Cuál será la resistencia si mezclo 390 kg de cemento, 570 kg de arena, 170 kg de agua y 1080 kg de agregado?

El resultado seria:
211.63 MPa con respecto a los resultados obtenidos, ya que con respecto de -1.66363793,  2.96670803,  0.63903057,  0.35848138,  2.31310047. Ya que con estos coeficientes, realizamos la formula correspondiente y lo sustituimos en lo dicho por la prgunta, y nos da **211.63 MPa**

## 4. Regresión logística

Un equipo de ingenieros quiere saber si una muestra de concreto hidráulico fallará o no bajo determinadas condiciones, basándose en experiencias pasadas.

<table>
  <thead>
    <tr>
      <th>Resistencia a la compresión (MPa)</th>
      <th>Edad del concreto (días)</th>
      <th>Relación agua-cemento</th>
      <th>Falla (Sí/No)</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>25</td>
      <td>28</td>
      <td>0.45</td>
      <td>Sí</td>
    </tr>
    <tr>
      <td>40</td>
      <td>56</td>
      <td>0.35</td>
      <td>No</td>
    </tr>
    <tr>
      <td>30</td>
      <td>28</td>
      <td>0.50</td>
      <td>Sí</td>
    </tr>
    <tr>
      <td>50</td>
      <td>90</td>
      <td>0.30</td>
      <td>No</td>
    </tr>
    <tr>
      <td>35</td>
      <td>28</td>
      <td>0.40</td>
      <td>Sí</td>
    </tr>
    <tr>
      <td>55</td>
      <td>120</td>
      <td>0.32</td>
      <td>No</td>
    </tr>
    <tr>
      <td>45</td>
      <td>60</td>
      <td>0.36</td>
      <td>No</td>
    </tr>
    <tr>
      <td>20</td>
      <td>7</td>
      <td>0.55</td>
      <td>Sí</td>
    </tr>
  </tbody>
</table>


1. ¿Qué factor afecta más? (Utiliza un vector de ceros como tu beta inicial)

In [7]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

data = {
    'Resistencia': [25, 40, 30, 50, 35, 55, 45, 20],
    'Edad': [28, 56, 28, 90, 28, 120, 60, 7],
    'Relación_agua_cemento': [0.45, 0.35, 0.50, 0.30, 0.40, 0.32, 0.36, 0.55],
    'Falla': ['Si', 'No', 'Si', 'No', 'Si', 'No', 'No', 'Si']
}

df = pd.DataFrame(data)

df['Falla'] = df['Falla'].map({'Si': 1, 'No': 0})

X = df[['Resistencia', 'Edad', 'Relación_agua_cemento']]
y = df['Falla']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

model = LogisticRegression()
model.fit(X_train, y_train)

y_pred = model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)
conf_matrix = confusion_matrix(y_test, y_pred)
classification_report_str = classification_report(y_test, y_pred)
coeficientes = model.coef_[0]
feature_names = X.columns
coef_df = pd.DataFrame(coeficientes, index=feature_names, columns=['Coeficiente'])

conf_matrix, classification_report_str, coef_df


/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:409: UserWarning: A single label was found in 'y_true' and 'y_pred'. For the confusion matrix to have the correct shape, use the 'labels' parameter to pass all known labels.
  warnings.warn(


(array([[2]]),
 '              precision    recall  f1-score   support\n\n           0       1.00      1.00      1.00         2\n\n    accuracy                           1.00         2\n   macro avg       1.00      1.00      1.00         2\nweighted avg       1.00      1.00      1.00         2\n',
                        Coeficiente
 Resistencia              -0.115218
 Edad                     -0.279618
 Relación_agua_cemento     0.000675)

2. Calcula la probabilidad de falla para una muestra con las siguientes características:
- Resistencia a la compresión: 38 MPa
- Edad del concreto: 30 días
- Relación agua-cemento: 0.40

In [8]:
nueva_muestra = pd.DataFrame({
    'Resistencia': [38],
    'Edad': [30],
    'Relación_agua_cemento': [0.40]
})

probabilidad_falla = model.predict_proba(nueva_muestra)[:, 1]
probabilidad_falla

array([0.98859978])